src/main/kotlin/edu/kit/compiler/Compiler.kt

```diff
@@ -130,42 +138,36 @@ class Compiler(private val config: Config) {
      * Sanity-check the input parameter and then prepare a reader that can be used by the lexer to generate a token
      * stream
      */
-    private fun openCompilationUnit(): CompilerResult<InputProvider> {
+    private fun openCompilationUnit(): CompilerResult<SourceFile> {
         val sourceFile = config.sourceFile
 
-        try {
-            if (!sourceFile.exists()) {
-                return CompilerResult.failure("File does not exist")
+        return try {
+            if (!Files.exists(sourceFile)) {
+                return CompilerResult.failure("source file does not exist")
             }
 
-            if (sourceFile.isDirectory) {
-                return CompilerResult.failure("Input path is not a file")
+            if (!Files.isRegularFile(sourceFile)) {
+                return CompilerResult.failure("source file is not a file")
             }
 
-            if (!sourceFile.canRead()) {
-                return CompilerResult.failure("Cannot read input file")
-            }
+            CompilerResult.success(
```

Comment:
This has a race condition with file deletion / permission settings etc anyways, so the checks above are just to provide slightly messages. You still need to handle all errors when trying to read from the file and *that* is where your "is regular file" check should be: implicitly done by the standardlib. If you use NIO.2, the error messages are useful and descriptive and cover the cases above nicely.

Reply:
same. Thanks :)